domain/schema_validator.go

```diff
@@ -41,91 +41,85 @@ type SchemaValidator interface {
 	return &schemaValidator{
+		isStarted:      true,
 		lease:          lease,
-		validItems:     make(map[int64]time.Time),
-		cacheItemInfos: make(map[int64]*deltaSchemaInfo),
+		itemSchemaVers: make(map[int64]struct{}),
+		detalItemInfos: make([]*deltaSchemaInfo, 0, maxNumberOfDiffsToLoad),
 	}
 }
 
 func (s *schemaValidator) Stop() {
 	log.Info("the schema validator stops")
 	s.mux.Lock()
 	defer s.mux.Unlock()
-	s.cacheItemInfos = nil
-	s.validItems = nil
+	s.isStarted = false
 	s.latestSchemaVer = 0
-	s.latestSchemaInfo = nil
+	s.detalItemInfos = nil
+	s.itemSchemaVers = nil
 }
 
 func (s *schemaValidator) Restart() {
 	log.Info("the schema validator restarts")
 	s.mux.Lock()
 	defer s.mux.Unlock()
-	s.validItems = make(map[int64]time.Time)
-	s.cacheItemInfos = make(map[int64]*deltaSchemaInfo)
+	s.isStarted = true
+	s.itemSchemaVers = make(map[int64]struct{})
+	s.detalItemInfos = make([]*deltaSchemaInfo, 0, maxNumberOfDiffsToLoad)
 }
 
 func (s *schemaValidator) Update(leaseGrantTS uint64, oldVer, currVer int64, changedTableIDs []int64) {
 	s.mux.Lock()
+	defer s.mux.Unlock()
 
-	if s.cacheItemInfos == nil {
-		s.mux.Unlock()
+	if !s.isStarted {
 		log.Infof("the schema validator stopped before updating")
 		return
 	}
 
 	// Renew the lease.
+	s.latestSchemaVer = currVer
 	leaseGrantTime := extractPhysicalTime(leaseGrantTS)
 	leaseExpire := leaseGrantTime.Add(s.lease - time.Millisecond)
-	s.validItems[currVer] = leaseExpire
-	_, hasCurrVerItemInfo := s.cacheItemInfos[currVer]
-	// Update the schema information.
-	if currVer != oldVer || currVer == 0 || !hasCurrVerItemInfo {
-		s.cacheItemInfos[currVer] = &deltaSchemaInfo{
-			expire:          leaseExpire,
-			relatedTableIDs: changedTableIDs,
-		}
-		oldInfo, ok := s.cacheItemInfos[oldVer]
-		if ok {
-			oldInfo.nextSchemaVersion = currVer
-		}
-	} else {
-		s.cacheItemInfos[currVer].expire = leaseExpire
+	s.latestSchemaExpire = leaseExpire
+
+	// Update the schema information map and slice.
+	_, hasCurrVerItemInfo := s.itemSchemaVers[currVer]
+	if currVer != oldVer || !hasCurrVerItemInfo {
+		s.itemSchemaVers[currVer] = struct{}{}
+		s.detalItemInfos = append(s.detalItemInfos, &deltaSchemaInfo{schemaVersion: currVer, relatedTableIDs: changedTableIDs})
 	}
-	s.latestSchemaVer = currVer
-	s.latestSchemaInfo = s.cacheItemInfos[currVer]
-
-	// Delete expired cacheItemInfos, leaseGrantTime is server current time, actually.
-	for k, info := range s.cacheItemInfos {
-		if leaseGrantTime.After(info.expire) {
-			delete(s.validItems, k)
-			// We cache some expired schema versions to store recently updated table IDs.
-			if shouldUpdateAllSchema(currVer, k) {
-				delete(s.cacheItemInfos, k)
-			}
+
+	// We cache some schema versions to store recently updated table IDs.
+	// If the schema version is much older than the latest schema version, we will delete it from itemSchemaVers.
+	offset := -1
+	for i, info := range s.detalItemInfos {
+		if isTooOldSchema(info.schemaVersion, currVer) {
```

Comment:
you should break the loop once you failed the `isTooOldSchema(info.schemaVersion, currVer)` branch, add `else { break }` ?